src/algebra/module/basic.lean

```diff
@@ -324,76 +332,53 @@ section add_comm_group
 
 variables [semiring S] [ring R] [add_comm_group M] [semimodule S M] [semimodule R M]
 
-/-- The natural ℤ-module structure on any `add_comm_group`. -/
--- We don't immediately make this a global instance, as it results in too many instances,
--- and confusing ambiguity in the notation `n • x` when `n : ℤ`.
--- We do turn it into a global instance, but only at the end of this file,
--- and I remain dubious whether this is a good idea.
-def add_comm_group.int_module : module ℤ M :=
-{ smul := gsmul,
-  smul_add := λ _ _ _, gsmul_add _ _ _,
-  add_smul := λ _ _ _, add_gsmul _ _ _,
-  mul_smul := λ _ _ _, gsmul_mul _ _ _,
-  one_smul := one_gsmul,
-  zero_smul := zero_gsmul,
-  smul_zero := gsmul_zero }
-
-section
-local attribute [instance] add_comm_group.int_module
-/-- `gsmul` is defined as the `smul` action of `add_comm_group.int_module`. -/
-lemma gsmul_def (n : ℤ) (x : M) : gsmul n x = n • x := rfl
-end
-
 section
 variables (R)
 /-- `gsmul` is equal to any other module structure via a cast. -/
-lemma gsmul_eq_smul_cast (n : ℤ) (b : M) : gsmul n b = (n : R) • b :=
+lemma gsmul_eq_smul_cast (n : ℤ) (b : M) : n • b = (n : R) • b :=
 begin
-  rw gsmul_def,
   induction n using int.induction_on with p hp n hn,
   { rw [int.cast_zero, zero_smul, zero_smul] },
   { rw [int.cast_add, int.cast_one, add_smul, add_smul, one_smul, one_smul, hp] },
   { rw [int.cast_sub, int.cast_one, sub_smul, sub_smul, one_smul, one_smul, hn] },
 end
 end
 
-/-- `gsmul` is equal to any `ℤ`-module structure. -/
-lemma gsmul_eq_smul [semimodule ℤ M] (n : ℤ) (b : M) : n •ℤ b = n • b :=
-by rw [gsmul_eq_smul_cast ℤ, n.cast_id]
+/-- Convert back any exotic `ℤ`-smul to the canonical instance. This should not be needed since in
+mathlib all `add_comm_group`s should normally have exactly one `ℤ`-module structure by design. -/
+lemma int_smul_eq_gsmul (h : semimodule ℤ M) (n : ℤ) (x : M) :
+  @has_scalar.smul ℤ M h.to_has_scalar n x = n • x :=
+by rw [gsmul_eq_smul_cast ℤ n x, int.cast_id]
 
-/-- All `ℤ`-module structures are equal. -/
-instance add_comm_group.int_module.subsingleton : subsingleton (semimodule ℤ M) :=
-⟨λ P Q, by {
-  ext n,
-  rw [←gsmul_eq_smul, ←gsmul_eq_smul], }⟩
+/-- All `ℤ`-module structures are equal. Not an instance since in mathlib all `add_comm_group`
+should normally have exactly one `ℤ`-module structure by design. -/
+def add_comm_group.int_module.unique : unique (semimodule ℤ M) :=
```

Comment:
Was the same `def` for `nat` already restored?

Reply:
Yes, in #7263. I copied the exact same formulation for int.

Reply:
Ok, thanks!